Asset: TerraformAssets are being redeployed without code changes

<!--- Please keep this note for the community --->

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### cdktf & Language Versions

- 0.4+
<!--- Please make note of the version you are using for the `hashicorp/terraform-cdk` and language bindings you are using.
If you are not running the latest version, please upgrade because your issue may have already been fixed. --->

### Affected Resource(s)

<!--- Please list the affected resources. --->
- TerraformAsset

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.
--->

```
[2021-08-02T11:50:40.461] [ERROR] cli-server - 2021-08-02T11:50:40.461+0200 [WARN]  Provider "provider[\"registry.terraform.io/hashicorp/aws\"]" produced an unexpected new value for aws_lambda_function.my-lambda-function, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .source_code_hash: was cty.StringVal("7F07D5FC474F00A5C1CD63C805926012"), but now cty.StringVal("Xr9qdAhR/qa5SNiNLyUJLWpMpo2+olazaKjh+Ld0KKI=")
```

### Expected Behavior

<!--- What should have happened? --->

When the referenced code of a `TerraformAsset` wasn't changed, there shouldn't be a change in the Terraform plan

### Actual Behavior

Code referenced via TerraformAsset and the `assetHash` property is always being re-deployed

--- a/test/typescript/asset/main.ts
+++ b/test/typescript/asset/main.ts
@@ -8,25 +8,37 @@ import {
   AssetType,
 } from "cdktf";
 import * as path from "path";
+import { Resource as NullResource, NullProvider } from "./.gen/providers/null";
 
-export class HelloTerra extends TerraformStack {
+function assets(scope: Construct, assetOverrides = {}) {
+  const localAsset = new TerraformAsset(scope, "local-asset", {
+    path: path.resolve(__dirname, "local-asset.txt"),
+    ...assetOverrides,
+  });
+
+  const fixtures = new TerraformAsset(scope, "fixtures", {
+    path: path.resolve(__dirname, "fixtures"),
+    ...assetOverrides,
+  });
+
+  const zippedFixtures = new TerraformAsset(scope, "zipped-fixtures", {
+    path: path.resolve(__dirname, "fixtures"),
+    ...assetOverrides,
+    type: AssetType.ARCHIVE,
+  });
+
+  return {
+    localAsset,
+    fixtures,
+    zippedFixtures,
+  };
+}
+
+export class FixedHash extends TerraformStack {
   constructor(scope: Construct, id: string) {
     super(scope, id);
-
-    const localAsset = new TerraformAsset(this, "local-asset", {
-      path: path.resolve(__dirname, "local-asset.txt"),
+    const { localAsset, fixtures, zippedFixtures } = assets(this, {
       assetHash: "hash",
-    });
-
-    const fixtures = new TerraformAsset(this, "fixtures", {
-      path: path.resolve(__dirname, "fixtures"),
-      assetHash: "hash",
-    });
-
-    const zippedFixtures = new TerraformAsset(this, "zipped-fixtures", {
-      path: path.resolve(__dirname, "fixtures"),
-      assetHash: "hash",
-      type: AssetType.ARCHIVE,
     });
 
     new TerraformOutput(this, "local-asset-name", {
@@ -43,6 +55,27 @@ export class HelloTerra extends TerraformStack {
   }
 }
 
+export class NormalHash extends TerraformStack {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    new NullProvider(this, "null", {});
+    const { localAsset, fixtures, zippedFixtures } = assets(this);
+
+    new NullResource(this, "resource", {
+      triggers: {
+        localAssetAssetHash: localAsset.assetHash,
+        localAssetPath: localAsset.path,
+        fixturesAssetHash: fixtures.assetHash,
+        fixturesPath: fixtures.path,
+        zippedFixturesAssetHash: zippedFixtures.assetHash,
+        zippedFixturesPath: zippedFixtures.path,
+      },
+    });
+  }
+}
+
 const app = Testing.stubVersion(new App({ stackTraces: false }));
-new HelloTerra(app, "stack");
+new FixedHash(app, "fixed");
+new NormalHash(app, "normal");
 app.synth();

--- a/test/typescript/asset/test.ts
+++ b/test/typescript/asset/test.ts
@@ -13,16 +13,16 @@ describe("full integration test", () => {
   });
 
   test("synth", async () => {
-    await driver.synth();
-    expect(driver.synthesizedStack("stack")).toMatchSnapshot();
+    await driver.synth("fixed");
+    expect(driver.synthesizedStack("fixed")).toMatchSnapshot();
   });
 
   test("file asset copied", async () => {
-    await driver.synth();
+    await driver.synth("fixed");
     expect(
       fs.readFileSync(
         path.resolve(
-          driver.stackDirectory("stack"),
+          driver.stackDirectory("fixed"),
           "assets/localasset/hash/local-asset.txt"
         ),
         "utf-8"
@@ -31,11 +31,11 @@ describe("full integration test", () => {
   });
 
   test("folder asset copied", async () => {
-    await driver.synth();
+    await driver.synth("fixed");
     expect(
       fs.readFileSync(
         path.resolve(
-          driver.stackDirectory("stack"),
+          driver.stackDirectory("fixed"),
           "assets/fixtures/hash/a.txt"
         ),
         "utf-8"
@@ -44,7 +44,7 @@ describe("full integration test", () => {
     expect(
       fs.readFileSync(
         path.resolve(
-          driver.stackDirectory("stack"),
+          driver.stackDirectory("fixed"),
           "assets/fixtures/hash/b.txt"
         ),
         "utf-8"
@@ -53,7 +53,7 @@ describe("full integration test", () => {
     expect(
       fs.readFileSync(
         path.resolve(
-          driver.stackDirectory("stack"),
+          driver.stackDirectory("fixed"),
           "assets/fixtures/hash/foo/bar/c.txt"
         ),
         "utf-8"
@@ -62,14 +62,24 @@ describe("full integration test", () => {
   });
 
   test("zip file created", async () => {
-    await driver.synth();
+    await driver.synth("fixed");
 
     const stat = fs.statSync(
       path.resolve(
-        driver.stackDirectory("stack"),
+        driver.stackDirectory("fixed"),
         "assets/zippedfixtures/hash/archive.zip"
       )
     );
     expect(stat.isFile()).toBe(true);
+  });
+
+  test("without asset changes there should be no redeployment", async () => {
+    expect(await driver.diff("normal")).toContain(
+      "1 to create, 0 to update, 0 to delete"
+    );
+    await driver.deploy("normal");
+    expect(await driver.diff("normal")).toContain(
+      "0 to create, 0 to update, 0 to delete"
+    );
   });
 });

